components/server/src/user/env-var-service.ts

```diff
@@ -0,0 +1,256 @@
+        return this.projectDB.setProjectEnvironmentVariable(projectId, name, value, censored);
+    }
+
+    async deleteProjectEnvironmentVariable(userId: string, variableId: string): Promise<void> {
+        const variable = await this.getProjectEnvironmentVariableById(userId, variableId);
+        await this.auth.checkPermissionOnProject(userId, "write_info", variable.projectId);
+        return this.projectDB.deleteProjectEnvironmentVariable(variableId);
+    }
+
+    async resolveEnvVariables(
+        userId: string,
+        projectId: string | undefined,
+        wsType: WorkspaceType,
+        wsContext: WorkspaceContext,
+    ): Promise<ResolvedEnvVars> {
+        const workspaceEnvVars = new Map<String, EnvVar>();
+        const merge = (envs: EnvVar[]) => {
+            for (const env of envs) {
+                workspaceEnvVars.set(env.name, env);
+            }
+        };
+
+        const projectEnvVars = projectId
+            ? (await ApplicationError.notFoundToUndefined(this.getProjectEnvironmentVariables(userId, projectId))) || []
+            : [];
+
+        if (wsType === "prebuild") {
+            // prebuild does not have access to user env vars and cannot be started via prewfix URL
+            const withValues = await this.projectDB.getProjectEnvironmentVariableValues(projectEnvVars);
+            merge(withValues);
+            return {
+                project: projectEnvVars,
+                workspace: [...workspaceEnvVars.values()],
+            };
+        }
+
+        // 1. first merge user envs
+        if (CommitContext.is(wsContext)) {
+            // this is a commit context, thus we can filter the env vars
+            const userEnvVars = await this.getAllEnvVars(userId, userId, {
+                canAccess: async () => {
+                    return true;
+                },
+            });
+            merge(UserEnvVar.filter(userEnvVars, wsContext.repository.owner, wsContext.repository.name));
+        }
+
+        // 2. then from the project
+        if (projectEnvVars.length) {
+            // Instead of using an access guard for Project environment variables, we let Project owners decide whether
+            // a variable should be:
+            //   - exposed in all workspaces (even for non-Project members when the repository is public), or
+            //   - censored from all workspaces (even for Project members)
+            const availablePrjEnvVars = projectEnvVars.filter((variable) => !variable.censored);
+            const withValues = await this.projectDB.getProjectEnvironmentVariableValues(availablePrjEnvVars);
+            merge(withValues);
+        }
+
+        // 3. then parsed from the context URL
+        if (WithEnvvarsContext.is(wsContext)) {
+            merge(wsContext.envvars);
+        }
+
+        return {
+            project: projectEnvVars,
+            workspace: [...workspaceEnvVars.values()],
+        };
+    }
+
+    private async guardAccess(
```

Comment:
maybe mark with @deprecated tag with a comment to remove after FGA migration

Reply:
we should not introduce it here